apiserver/root.go

```diff
@@ -712,38 +710,46 @@ func (r *apiHandler) ConnectedModel() string {
 }
 
-func (r *apiHandler) userPermission(subject names.UserTag, target names.Tag) (permission.Access, error) {
-	if r.authToken == nil {
-		return r.state.UserPermission(subject, target)
-	}
-	return permissionFromToken(r.authToken, target)
-}
-
-type entityHasPermissionFunc func(entity names.Tag, operation permission.Access, target names.Tag) (bool, error)
-
-// HasPermission returns true if the logged in user can perform <operation> on <target>.
-// If a login token is used to specify user and access, and the operaton is not allowed, an
-// AccessRequiredError is returned with the required permissions.
+// HasPermission returns true if the logged in user can perform <operation> on
+// <target>. If the logged in user does not have permission then false is
+// returned with a permission error specific to the users authentication
+// mechanism. The error will also satisfy Is(err, ErrorEntityMissingPermission).
 func (r *apiHandler) HasPermission(operation permission.Access, target names.Tag) (bool, error) {
-	return r.EntityHasPermission(r.entity.Tag(), operation, target)
+	return r.EntityHasPermission(r.GetAuthTag(), operation, target)
 }
 
-// EntityHasPermission returns true if the passed in entity can perform <operation> on <target>.
+// EntityHasPermission returns true if the passed in entity can perform
+// <operation> on <target>. If entity user does not have permission then
+// false is returned with a permission error specific to the users
+// authentication mechanism. The error will also satisfy
+// Is(err, ErrorEntityMissingPermission).
 func (r *apiHandler) EntityHasPermission(entity names.Tag, operation permission.Access, target names.Tag) (bool, error) {
-	has, err := common.HasPermission(r.userPermission, entity, operation, target)
-	if r.authToken == nil || err != nil || has {
-		return has, err
+	var userAccessFunc common.UserAccessFunc = func(entity names.UserTag, subject names.Tag) (permission.Access, error) {
+		if r.authInfo.Delegator == nil {
+			return permission.NoAccess, fmt.Errorf("permissions %w for auth info", errors.NotImplemented)
+		}
+		return r.authInfo.Delegator.SubjectPermissions(authentication.TagToEntity(entity), subject)
 	}
-	return false, &apiservererrors.AccessRequiredError{
-		RequiredAccess: map[names.Tag]permission.Access{
-			target: operation,
-		},
+	has, err := common.HasPermission(userAccessFunc, entity, operation, target)
+	if err != nil && !errors.Is(err, errors.NotFound) {
+		return false, fmt.Errorf("checking entity %q has permission: %w", entity, err)
 	}
+	if !has && r.authInfo.Delegator != nil {
+		err = r.authInfo.Delegator.PermissionError(target, operation)
+	}
+	if !has {
+		return false, errors.WithType(err, authentication.ErrorEnityMissingPermission)
+	}
+
+	return true, nil
 }
 
 // AuthTokenString returns the jwt passed to login.
 func (r *apiHandler) AuthTokenString() string {
-	return r.authTokenString
+	// TODO fix up in auth code
+	// tlm
+	return ""
+	// return r.authTokenString
```

Comment:
Fix me

Reply:
Done, we are going to re address this later with follow up work for JAAS. It is currently not needed.